melusine/models/models_v2/base_model.py

```diff
@@ -0,0 +1,175 @@
+        """
+        getstate is called when a class is pickled.
+        Class attributes that should not be saved can be specified in the pkl_exclude list.
+        For example, a pretrained embedding attribute should not be saved.
+
+        Returns
+        -------
+        _: dict
+            Instance dict to be saved
+        """
+        instance_dict = self.__dict__.copy()
+        for key in self.EXCLUDE_LIST:
+            instance_dict.pop(key)
+        return instance_dict
+
+    def fit_label_encoder(self, y: pd.Series) -> None:
+        """
+        Encode categorical labels.
+
+        Parameters
+        ----------
+        y: pd.Series
+            Labels
+        """
+        le = LabelEncoder()
+        le.fit(y.astype(str))
+        self.label_encoder = le
+        self.n_targets = len(self.label_encoder.classes_)
+
+        logger.info(f"Finished fitting label encoder ({self.n_targets} classes)")
+
+    def fit_meta_cols(self, x: pd.DataFrame) -> None:
+
+        if not self.meta_input_list:
+            meta_cols = None
+        else:
+            meta_prefix_list = tuple(f"{col}__" for col in self.meta_input_list)
+            meta_cols = [col for col in x.columns if col.startswith(meta_prefix_list)]
+
+        self.meta_cols = meta_cols
+        logger.info("Finished fitting meta columns")
+        logger.info(f" Meta columns list : {self.meta_cols}")
+
+    def get_meta_input(
+        self, inputs: List[np.ndarray], x: pd.DataFrame
+    ) -> List[np.ndarray]:
+
+        if not self.meta_cols:
+            return inputs
+
+        inputs.append(x[self.meta_cols].values)
+        return inputs
+
+    def add_meta_net(
+        self, inputs: List[np.ndarray], net, meta_archi: Callable
+    ) -> (List[np.ndarray], str):  # TODO change
+
+        # Meta features net
+        if self.meta_cols:
+            n_meta_features = len(self.meta_cols)
+            meta_input, meta_net = meta_archi(n_meta_features)
+            inputs.append(meta_input)
+            net = Concatenate(axis=1)([net, meta_net])
+            logger.info(f"Added meta net ({n_meta_features} meta features)")
+        else:
+            net = net
+            logger.info(f"No meta features found, skipping meta net")
+
+        return inputs, net
+
+    @abstractmethod
+    def fit(self, x: pd.DataFrame, y: pd.Series = None) -> None:
+        """
+
+        Parameters
+        ----------
+        x: pandas.DataFrame
+            DataFrame with raw features
+        y: pandas.Series
+            Series with raw categories
+        """
+        self.fit_meta_cols(x)
+        self.fit_label_encoder(y)
+
+    @abstractmethod
+    def transform(self, x: pd.DataFrame) -> List[np.ndarray]:
+        pass
+
+    def encode_labels(self, y: pd.Series) -> np.ndarray:
+        return to_categorical(self.label_encoder.transform(y))
+
+    def save_model(self, path: str) -> None:
+        self.model.save(path)
+        logger.info(f"Saved Deep Learning model to {path}")
+
```

Comment:
how about a MLFLOW Connector ? With the equivalent of [autolog](https://www.mlflow.org/docs/latest/python_api/mlflow.sklearn.html) method for sklearn ?
Maybe too soon and out of scope for now but maybe not so complicated

Reply:
I think/hope this could be achieved through the MelusineTrainer class.
PyTorch / HuggingFace trainers make it easy to link model training to MlFlow, I'll have a look at it.

On top of that I think the Trainer should implement EarlyStopping, Logging and Tensorboard callbacks.
(Will give it a try soon)